README.adoc
= Go-Wardley - Wardley maps generator
David Gamba, https://github.com/DavidGamba
:idprefix:
:toc: macro

Wardley Maps Generator written in Golang.
The generator takes an HCL based map description and generates a map in SVG format.
If viewed on a browser, the SVG will have hover over functionality to get additional details on node descriptions.

toc::[]

== Usage

----
$ ./go-wardley -f examples/map.hcl
Updated file: examples/map.svg

$ ./go-wardley -f examples/map.hcl -o examples/map.svg
Updated file: examples/map.svg

# Watch for file changes and update the file automatically.
$ ./go-wardley -f examples/map.hcl --watch
Starting watcher on: examples
Updated file: examples/map.svg
Updated file: examples/map.svg

# Serve the file on a webserver in localhost:8080 by default
# Update the drawing by refreshing the page.
$ ./go-wardley -f examples/map.hcl --serve
Serving content on: http://localhost:8080
$ ./go-wardley -f examples/map.hcl --serve 6060
Serving content on: http://localhost:6060
----

image::./examples/map.svg[]

== Element types

=== Size

----
size {
	width     = 1280
	height    = 768
	margin    = 40
	font_size = 9
}
----

=== Node

----
node user {
	label       = "User"        # Required
	visibility  = 1             # Required
	evolution   = "custom"      # Required
	x           = 1             # Required
	description = "Description"
	fill        = "black"
	color       = "black"
}
----

`evolution`:: `genesis`, `custom`, `product` or `commodity`.

=== Connector

----
connector {
	from  = "user"        # Required
	to    = "vcs"         # Required
	label = "Description"
	color = "black"
	type  = "normal"
}
----

`type`:: `normal`, `bold`, `change` or `change-inertia`.

== Example input

A more extensive example can be found in link:./examples/map.hcl[].

[source, hcl]
----
# Anchor
node user {
	label       = "User"
	visibility  = 1
	evolution   = "custom"
	x           = 1
	description = "User Description"
	fill        = "black"
}

node vcs {
	label       = "On Prem VCS"
	visibility  = node.user.visibility + 1
	evolution   = "product"
	x           = 1
	description = "On prem VCS"
	fill        = "black"
}

node code_commit {
	label       = "Code Commit Mirror"
	visibility  = node.vcs.visibility
	evolution   = "commodity"
	x           = 1
	description = "Allows Code Pipeline to access the code."
	color       = "red"
}

connector {
	from = "user"
	to   = "vcs"
}

connector {
	from = "vcs"
	to   = "code_commit"
	color = "red"
	type = "change-inertia"
}
----

== Roadmap

* Make the node label optional, read the node ID if not present and title case it (configurable?).

* Add an optional leyend box indicating the types of elements used in the map.

* Better looks overall. Cleaner code.

* Allow specifying node, connector and grid font sizes independently.

* Arch type connector.

== License

This file is part of go-wardley.

Copyright (C) 2019-2020  David Gamba Rios

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at http://mozilla.org/MPL/2.0/.


changelog.adoc
= Changelog

== v0.3.0

* Update HCL code to allow element references.
It allows to reference a previous element in the next one.
* Update README.
* Update example.
* https://github.com/DavidGamba/go-wardley/pull/1[Fix Calculation of Node X Position Relative to Max x For Evolution], thanks https://github.com/cwarden[@cwarden]!

== v0.2.0

* Add a changelog.
* Add option to print executable version: `./go-wardley -V` or `./go-wardley --version`.
* Update HCL parsing code to allow for optional values.
* Update HCL parsing to print user facing diagnostic errors.
* Add fsnotify file watches.
* Allow defining Map Details (size) in HCL.

== v0.1.0

Sharing MVP with the world, it sucks but it works.


hcl/hcl_test.go
// This file is part of go-wardley.
//
// Copyright (C) 2019-2020  David Gamba Rios
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

package hcl

import (
	"bytes"
	"reflect"
	"testing"

	"github.com/davecgh/go-spew/spew"
)

func TestDecodeMap(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected *Map
	}{
		{"empty", "", &mapDefaults},
		{"empty", "size {}", &mapDefaults},
		{"optional", "size { width = 7 }", &Map{
			Size: &Size{Width: 7, Height: 768, Margin: 40, FontSize: 12},
		}},
		{"node", `node id {
				label = "label"
				visibility = 1
				evolution = "custom"
				x = 1
			}`, &Map{
			Size:  &Size{Width: 1280, Height: 768, Margin: 40, FontSize: 12},
			Nodes: []*Node{{ID: "id", Label: "label", Visibility: 1, Evolution: "custom", EvolutionX: 1, Fill: "white", Color: "black"}},
		}},
		{"connector", `connector {
				label = "label"
				to = "to"
				from = "from"
			}`, &Map{
			Size:       &Size{Width: 1280, Height: 768, Margin: 40, FontSize: 12},
			Connectors: []*Connector{{Label: "label", To: "to", From: "from", Color: "black", Type: "normal"}},
		}},
		{"all", `node id {
				label = "label"
				visibility = 1
				evolution = "custom"
				x = 1
			}
			node id2 {
				label = "label"
				visibility = 1
				evolution = "custom"
				x = 1
			}
			connector {
				label = "label"
				to = "to"
				from = "from"
			}`, &Map{
			Size: &Size{Width: 1280, Height: 768, Margin: 40, FontSize: 12},
			Nodes: []*Node{
				{ID: "id", Label: "label", Visibility: 1, Evolution: "custom", EvolutionX: 1, Fill: "white", Color: "black"},
				{ID: "id2", Label: "label", Visibility: 1, Evolution: "custom", EvolutionX: 1, Fill: "white", Color: "black"},
			},
			Connectors: []*Connector{{Label: "label", To: "to", From: "from", Color: "black", Type: "normal"}},
		}},
		{"references", `node id {
				label = "label"
				visibility = 1
				evolution = "custom"
				x = 1
			}
			node id2 {
				label = "label"
				visibility = node.id.visibility
				evolution = "custom"
				x = node.id.x + 1
			}
			node id3 {
				label = "label"
				visibility = node.id.visibility
				evolution = "custom"
				x = node.id2.x + 1
			}
			connector {
				label = "label"
				to = "to"
				from = "from"
			}`, &Map{
			Size: &Size{Width: 1280, Height: 768, Margin: 40, FontSize: 12},
			Nodes: []*Node{
				{ID: "id", Label: "label", Visibility: 1, Evolution: "custom", EvolutionX: 1, Fill: "white", Color: "black"},
				{ID: "id2", Label: "label", Visibility: 1, Evolution: "custom", EvolutionX: 2, Fill: "white", Color: "black"},
				{ID: "id3", Label: "label", Visibility: 1, Evolution: "custom", EvolutionX: 3, Fill: "white", Color: "black"},
			},
			Connectors: []*Connector{{Label: "label", To: "to", From: "from", Color: "black", Type: "normal"}},
		}},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			buf := new(bytes.Buffer)
			parser, f, err := ParseHCL(buf, []byte(test.input), "test.hcl")
			if err != nil {
				t.Fatalf("%s\n%s\n", err, buf.String())
			}
			mapDetails, err := DecodeMap(buf, parser, f)
			if buf.String() != "" {
				t.Errorf("output didn't match expected value: %s", buf.String())
			}
			if err != nil {
				out := new(bytes.Buffer)
				spew.Fdump(out, mapDetails)
				t.Fatalf("Error: %s, %s", err, out.String())
			}
			if !reflect.DeepEqual(mapDetails, test.expected) {
				out := new(bytes.Buffer)
				exp := new(bytes.Buffer)
				spew.Fdump(out, mapDetails)
				spew.Fdump(exp, test.expected)
				t.Fatalf("unexpected value:\n%s!=\n%s", out.String(), exp.String())
			}
		})
	}
}


cue/cue.go
package main

import (
	"context"
	"embed"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"os"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	cueErrors "cuelang.org/go/cue/errors"
	"cuelang.org/go/encoding/gocode/gocodec"
	"github.com/DavidGamba/go-getoptions"
)

//go:embed wardley-schema.cue
var f embed.FS

var Logger = log.New(os.Stderr, "", log.LstdFlags)

func main() {
	os.Exit(program(os.Args))
}

func program(args []string) int {
	opt := getoptions.New()
	opt.SetUnknownMode(getoptions.Pass)
	opt.Bool("quiet", false, opt.GetEnv("QUIET"))
	opt.SetCommandFn(Run)
	opt.String("config", "", opt.Required())
	opt.HelpCommand("help", opt.Alias("?"))
	remaining, err := opt.Parse(args[1:])
	if err != nil {
		fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
		return 1
	}
	if opt.Called("quiet") {
		Logger.SetOutput(io.Discard)
	}
	Logger.Println(remaining)

	ctx, cancel, done := getoptions.InterruptContext()
	defer func() { cancel(); <-done }()

	err = opt.Dispatch(ctx, remaining)
	if err != nil {
		if errors.Is(err, getoptions.ErrorHelpCalled) {
			return 1
		}
		fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
		return 1
	}
	return 0
}

func Run(ctx context.Context, opt *getoptions.GetOpt, args []string) error {
	Logger.Printf("Parsing cue config")

	configFilename := opt.Value("config").(string)

	c := cuecontext.New()
	schema, err := f.ReadFile("wardley-schema.cue")
	if err != nil {
		return fmt.Errorf("failed to read schema: %w", err)
	}
	// compile our schema first
	s := c.CompileBytes(schema)

	configData, err := os.ReadFile(configFilename)
	if err != nil {
		return fmt.Errorf("failed to read config file: %w", err)
	}
	// compile our value with scope
	v := c.CompileBytes(configData, cue.Scope(s))
	if v.Err() != nil {
		return fmt.Errorf("failed to compile config: %s", cueErrors.Details(v.Err(), nil))
	}
	err = v.Validate(
		cue.Final(),
		cue.Concrete(true),
		cue.Definitions(true),
		cue.Hidden(true),
		cue.Optional(true),
	)
	if err != nil {
		// TODO: We could print the line with the error by parsing the possitions response.
		// Logger.Printf("possitions: %v\n", cueErrors.Positions(err))
		return fmt.Errorf("failed config validation of file '%s': %s", configFilename, cueErrors.Details(err, nil))
	}

	w := Wardley{}

	g := gocodec.New((*cue.Runtime)(c), nil)
	err = g.Encode(v, &w)
	if err != nil {
		return err
	}
	fmt.Printf("map: %v\n", w)
	pretty, err := json.MarshalIndent(w, "", "\t")
	if err != nil {
		return err
	}
	fmt.Printf("map: %v\n", string(pretty))

	i, err := v.Fields()
	if err != nil {
		return fmt.Errorf("failed to get fields: %w", err)
	}
	for i.Next() {
		fmt.Printf("value %s: %v\n", i.Value().Path(), i.Value())
	}

	return nil
}

type Wardley struct {
	Map struct {
		Size      Size
		Node      map[string]Node
		Connector map[string]Connector
	}
}

type Size struct {
	Width    int
	Height   int
	Margin   int
	FontSize int `json:"font_size"`
}

type Node struct {
	ID          string
	Label       string
	Visibility  int
	Evolution   string
	X           int
	Description string
	Fill        string
	Color       string
}

type Connector struct {
	ID    string
	From  string
	To    string
	Label string
	Color string
	Type  string
}


main.go
// This file is part of go-wardley.
//
// Copyright (C) 2019-2020  David Gamba Rios
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"math/rand"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/DavidGamba/go-getoptions"
	"github.com/DavidGamba/go-wardley/hcl"
	svg "github.com/ajstarks/svgo"
	"github.com/fsnotify/fsnotify"
)

// BuildMetadata - Provides the metadata part of the version information.
var BuildMetadata string = "dev"

var version string = "0.3.0"

var logger = log.New(ioutil.Discard, "", log.LstdFlags)

// Show guides in drawing
var showGuides bool

// Add jitter to nodes' X and Y positions to reduce overlapping
var jitter bool

// Keeps count of connect path IDs
var connectID int

var canvas *svg.SVG

func main() {
	var inputFile, outputFile string
	var port int

	opt := getoptions.New()
	opt.Bool("help", false, opt.Alias("?"))
	opt.Bool("debug", false, opt.Description("Show debug logs"))
	opt.BoolVar(&jitter, "jitter", false, opt.Description("Add jitter to node positions"))
	opt.IntVarOptional(&port, "serve", 8080, opt.Description("Serve the drawing at localhost:<port>"), opt.ArgName("port"))
	opt.Bool("version", false, opt.Alias("V"), opt.Description("Print version information"))
	opt.Bool("watch", false, opt.Description("Watch file for changes"))
	opt.BoolVar(&showGuides, "guides", false, opt.Description("Show margins, limits and other guides in drawing"))
	opt.StringVar(&inputFile, "file", "", opt.Description("Map input file"), opt.Required(""), opt.ArgName("filename"))
	opt.StringVar(&outputFile, "output", "", opt.Description("Map svg output file, by default replaces input file extension to .svg"), opt.ArgName("filename"))
	_, err := opt.Parse(os.Args[1:])
	if opt.Called("help") {
		fmt.Println(opt.Help())
		os.Exit(1)
	}
	if opt.Called("version") {
		fmt.Printf("Version: %s+%s\n", version, BuildMetadata)
		os.Exit(1)
	}
	if err != nil {
		fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
		os.Exit(1)
	}
	if opt.Called("debug") {
		logger.SetOutput(os.Stderr)
		hcl.Logger.SetOutput(os.Stderr)
	}

	if opt.Called("serve") {
		fmt.Printf("Serving content on: http://localhost:%d\n", port)
		serveFile(inputFile, port)
		os.Exit(0)
	}

	if opt.Called("watch") {
		absFile, err := filepath.Abs(inputFile)
		if err != nil {
			fmt.Fprintf(os.Stderr, "ERROR: failed to get Absolute path from input file '%s': %s\n", inputFile, err)
			os.Exit(1)
		}
		watcher, err := fsnotify.NewWatcher()
		if err != nil {
			fmt.Fprintf(os.Stderr, "ERROR: failed to stablish watcher: %s\n", err)
			os.Exit(1)
		}
		defer watcher.Close()

		done := make(chan bool)
		go func() {
			for {
				select {
				case event, ok := <-watcher.Events:
					if !ok {
						return
					}
					logger.Printf("watcher event: %s\n", event.String())
					if event.Name == absFile && event.Op&fsnotify.Write == fsnotify.Write {
						err := render(absFile, outputFile)
						if err != nil {
							fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
						}
					}
				case err, ok := <-watcher.Errors:
					if !ok {
						return
					}
					logger.Printf("Watcher error: %s\n", err)
				}
			}
		}()

		// Vim deletes the file when saving it making fsnotify loose the pointer to it.
		// Have to watch the dir.
		fmt.Printf("Starting watcher on: %s\n", filepath.Dir(absFile))
		err = watcher.Add(filepath.Dir(absFile))
		if err != nil {
			fmt.Fprintf(os.Stderr, "ERROR: watcher error: %s\n", err)
			os.Exit(1)
		}
		err = render(absFile, outputFile)
		if err != nil {
			fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
		}
		<-done
	} else {
		err := render(inputFile, outputFile)
		if err != nil {
			fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
			os.Exit(1)
		}
	}
}

func render(inputFile, outputFile string) error {
	m, err := parseInputFile(inputFile)
	if err != nil {
		return err
	}
	if outputFile == "" {
		outputFile = strings.Replace(inputFile, filepath.Ext(inputFile), ".svg", 1)
		logger.Printf("output file: %s\n", outputFile)
	}
	err = renderFile(m, outputFile)
	if err != nil {
		return err
	}
	return nil
}

func parseInputFile(name string) (*hcl.Map, error) {
	parser, f, err := hcl.ParseHCLFile(os.Stderr, name)
	if err != nil {
		return nil, fmt.Errorf("failed to read '%s': %w", name, err)
	}
	m, err := hcl.DecodeMap(os.Stderr, parser, f)
	if err != nil {
		return nil, err
	}
	return m, nil
}

func renderFile(m *hcl.Map, outputFile string) error {
	ofh, err := os.Create(outputFile)
	if err != nil {
		return fmt.Errorf("failed to write to '%s': %w", outputFile, err)
	}
	defer ofh.Close()
	drawing(ofh, m)
	fmt.Printf("Updated file: %s\n", outputFile)
	return nil
}

func serveFile(inputFile string, port int) error {
	http.Handle("/", http.HandlerFunc(drawHandler(inputFile)))
	err := http.ListenAndServe(fmt.Sprintf(":%d", port), nil)
	if err != nil {
		return err
	}
	return nil
}

func drawHandler(inputFile string) func(w http.ResponseWriter, req *http.Request) {
	return func(w http.ResponseWriter, req *http.Request) {
		m, err := parseInputFile(inputFile)
		if err != nil {
			fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
		}
		w.Header().Set("Content-Type", "image/svg+xml")
		drawing(w, m)
	}
}

func drawing(w io.Writer, m *hcl.Map) {
	canvas = svg.New(w)
	canvas.Start(m.Size.Width, m.Size.Height)
	canvas.Gstyle("font-family:sans-serif")
	grid(canvas, m.Size.Margin, m.Size.Width, m.Size.Height, m.Size.FontSize+2)
	canvas.Translate(m.Size.Margin*2, m.Size.Height-m.Size.Margin*2)
	canvas.Marker("connector-arrow", 17, 3, 12, 10, `orient="auto"`)
	canvas.Path("M0,0 L0,6 L12,3 z")
	canvas.MarkerEnd()
	canvas.Marker("connector-inertia", 0, 10, 20, 40, `orient="auto"`)
	canvas.Path("M-5,20 L-5,-20 L5,-20 L5,20")
	canvas.MarkerEnd()

	nodes := m.Nodes
	connectors := m.Connectors

	maxGenesis, maxCustom, maxProduct, maxCommodity := 0, 0, 0, 0
	maxY := 0
	for _, n := range nodes {
		if n.Evolution == "genesis" && n.EvolutionX > maxGenesis {
			maxGenesis = n.EvolutionX
		}
		if n.Evolution == "custom" && n.EvolutionX > maxCustom {
			maxCustom = n.EvolutionX
		}
		if n.Evolution == "product" && n.EvolutionX > maxProduct {
			maxProduct = n.EvolutionX
		}
		if n.Evolution == "commodity" && n.EvolutionX > maxCommodity {
			maxCommodity = n.EvolutionX
		}
		if n.Visibility > maxY {
			maxY = n.Visibility
		}
	}
	for _, n := range nodes {
		NodeXY(n, maxGenesis, maxCustom, maxProduct, maxCommodity, maxY)
	}
	for _, c := range connectors {
		var a, b *hcl.Node
		for _, n := range nodes {
			if n.ID == c.From {
				a = n
			}
			if n.ID == c.To {
				b = n
			}
		}
		if a == nil || b == nil {
			fmt.Fprintf(os.Stderr, "ERROR: couldn't find node '%s'\n", c.From)
			continue
		}
		if b == nil {
			fmt.Fprintf(os.Stderr, "ERROR: couldn't find node '%s'\n", c.To)
			continue
		}
		connect(c, a, b, m.Size.FontSize)
	}
	for _, n := range nodes {
		DrawNode(n, m.Size.FontSize)
	}
	canvas.Gend()
	canvas.Gend()
	canvas.End()
}

var mapGrid Grid

// Grid -
type Grid struct {
	XQuarterLength int
	YLength        int
	Genesis        int
	Custom         int
	Product        int
	Commodity      int
	Visible        int
}

func NodeXY(n *hcl.Node, maxGenesis, maxCustom, maxProduct, maxCommodity, maxY int) {
	switch n.Evolution {
	case "genesis":
		n.X = mapGrid.Genesis + mapGrid.XQuarterLength/(maxGenesis+1)*n.EvolutionX
	case "custom":
		n.X = mapGrid.Custom + mapGrid.XQuarterLength/(maxCustom+1)*n.EvolutionX
	case "product":
		n.X = mapGrid.Product + mapGrid.XQuarterLength/(maxProduct+1)*n.EvolutionX
	case "commodity":
		n.X = mapGrid.Commodity + mapGrid.XQuarterLength/(maxCommodity+1)*n.EvolutionX
	}

	n.Y = -mapGrid.YLength / (maxY + 1) * (maxY + 1 - n.Visibility)
	if jitter {
		jitterSizeX := mapGrid.XQuarterLength / 20
		jitterX := -jitterSizeX + rand.Intn(jitterSizeX*2+1)
		jitterSizeY := mapGrid.YLength / 100
		jitterY := -jitterSizeY + rand.Intn(jitterSizeY*2+1)
		n.X += jitterX
		n.Y += jitterY
	}
}

// DrawNode -
func DrawNode(n *hcl.Node, fontSize int) {
	canvas.Gstyle("text-shadow: 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white")
	if n.Description != "" {
		canvas.Title(n.Description)
	} else {
		canvas.Title(n.Label)
	}
	canvas.Circle(n.X, n.Y, 5, fmt.Sprintf("fill:%s;stroke:%s", n.Fill, n.Color))
	// canvas.Text(n.X+10, n.Y+3, n.Label, fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:black;text-shadow: -1px 0 white, 0 1px white, 1px 0 white, 0 -1px white", nodeFontSize))
	// canvas.Gstyle("text-shadow: -1px 0 white, 0 1px white, 1px 0 white, 0 -1px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white, 0 0 3px white")
	canvas.Textlines(n.X+8, n.Y+10, strings.Split(n.Label, "\n"), fontSize, fontSize+3, "black", "left")
	canvas.Gend()

}

func connect(c *hcl.Connector, a, b *hcl.Node, fontSize int) {
	connectID++

	// Calculate midpoints
	x := a.X + (b.X-a.X)/2
	if a.X > b.X {
		x = b.X + (a.X-b.X)/2
	}
	y := a.Y + (b.Y-a.Y)/2
	if a.Y > b.Y {
		y = b.Y + (a.Y-b.Y)/2
	}

	// canvas.Def()
	// canvas.Path(fmt.Sprintf("M %d,%d %d,%d %d,%d", a.X, a.Y, x, y, b.X, b.Y), fmt.Sprintf(`id="%d"`, connectID))
	// canvas.DefEnd()
	switch c.Type {
	case "normal":
		canvas.Path(fmt.Sprintf("M %d,%d %d,%d", a.X, a.Y, b.X, b.Y),
			fmt.Sprintf(`id="%s-%s"`, a.ID, b.ID),
			fmt.Sprintf(`fill:none;stroke:%s;opacity:0.2`, c.Color))
	case "bold":
		canvas.Path(fmt.Sprintf("M %d,%d %d,%d", a.X, a.Y, b.X, b.Y), fmt.Sprintf(`fill:none;stroke:%s;opacity:0.8`, c.Color))
	case "change":
		canvas.Path(fmt.Sprintf("M %d,%d %d,%d %d,%d", a.X, a.Y, x, y, b.X, b.Y),
			fmt.Sprintf(`fill:white;stroke:%[1]s;opacity:0.6;stroke-dasharray:6,6;marker-end:url(#connector-arrow)`, c.Color))
	case "change-inertia":
		canvas.Path(fmt.Sprintf("M %d,%d %d,%d %d,%d", a.X, a.Y, x, y, b.X, b.Y),
			fmt.Sprintf(`fill:white;stroke:%[1]s;opacity:0.6;stroke-dasharray:6,6;marker-mid:url(#connector-inertia);marker-end:url(#connector-arrow)`, c.Color))
	}
	x += 8
	y += 10

	// if strings.Contains(c.Label, "\n") {
	canvas.Textlines(x, y, strings.Split(c.Label, "\n"), fontSize, fontSize+3, "black", "left")
	// } else {
	// 	s.Textpath(c.Label, fmt.Sprintf("#%d", connectID), `x="10" y="-5"`, fmt.Sprintf("text-align:left;font-size:%dpx;fill:black", nodeFontSize))
	// }
}

func grid(s *svg.SVG, margin, width, height, fontSize int) {
	// Grid
	//   X
	xLength := width - margin*4
	xZero := margin * 2
	xEnd := width - margin*2
	//   Y
	yLength := height - margin*4
	yZero := height - margin*2
	yEnd := margin * 2

	mapGrid = Grid{
		XQuarterLength: (width - margin*4) / 4,
		Genesis:        0,
		Custom:         (width - margin*4) / 4,
		Product:        (width - margin*4) * 2 / 4,
		Commodity:      (width - margin*4) * 3 / 4,
		YLength:        height - margin*4,
		Visible:        0,
	}

	s.Rect(0, 0, width, height, "fill:white")

	if showGuides {
		// Limits Guide
		s.Line(0, height, width, height, "fill:none;stroke:red")
		s.Line(width, 0, width, height, "fill:none;stroke:red")
		// Margin guide
		s.Line(margin, height-margin, width-margin, height-margin, "fill:none;stroke:green")
		s.Line(width-margin, margin, width-margin, height-margin, "fill:none;stroke:green")
		s.Line(margin, margin, width-margin, margin, "fill:none;stroke:green")
		s.Line(margin, margin, margin, height-margin, "fill:none;stroke:green")

		s.Translate(xZero, yZero)
		s.Text(xLength-40, -yLength, fmt.Sprintf("%d,%d", xLength, yLength), fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:green", fontSize))
		s.Text(0, 0, fmt.Sprintf("%d,%d", xZero, yZero), fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:green", fontSize))
		s.Text(xLength/4, 0, fmt.Sprintf("%d,%d", xLength/4, 0), fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:green", fontSize))
		s.Text(xLength*2/4, 0, fmt.Sprintf("%d,%d", 2*margin+xLength*2/4, 0), fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:green", fontSize))
		s.Text(xLength*3/4, 0, fmt.Sprintf("%d,%d", 2*margin+xLength*3/4, 0), fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:green", fontSize))
		s.Gend()
	}

	// Grid
	s.Marker("arrow", 0, 3, 12, 10, `orient="auto"`)
	s.Path("M0,0 L0,6 L12,3 z", "fill:black")
	s.MarkerEnd()
	s.Line(xZero, yZero, xEnd, yZero, "fill:none;stroke:black;marker-end:url(#arrow)")
	s.Line(xZero, yZero, xZero, yEnd, "fill:blue;stroke:black;marker-end:url(#arrow)")

	s.Line(2*margin+xLength/4, yZero, 2*margin+xLength/4, yEnd, `fill:none;stroke:gray;stroke-dasharray:1,10`)
	s.Line(2*margin+xLength*2/4, yZero, 2*margin+xLength*2/4, yEnd, `fill:none;stroke:gray;stroke-dasharray:1,10`)
	s.Line(2*margin+xLength*3/4, yZero, 2*margin+xLength*3/4, yEnd, `fill:none;stroke:gray;stroke-dasharray:1,10`)

	// Text
	s.Text(xZero, height-margin, "Genesis", fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:black", fontSize))
	s.Text(2*margin+xLength/4, height-margin, "Custom", fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:black", fontSize))
	s.Text(2*margin+xLength*2/4, height-margin, "Product (+rental)", fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:black", fontSize))
	s.Text(2*margin+xLength*3/4, height-margin, "Commodity (+utility)", fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:black", fontSize))
	s.Text(xEnd-100, height-2*margin-5, "Evolution", fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:black;font-weight:bold;font-family:serif", fontSize+2))

	s.TranslateRotate(xZero, yZero, 270)
	s.Text(0, -5, "Invisible", fmt.Sprintf("text-anchor:top;font-size:%dpx;fill:black", fontSize))
	s.Text(yLength-50, -5, "Visible", fmt.Sprintf("text-anchor:top;font-size:%dpx;fill:black", fontSize))
	s.Text(yLength-100, fontSize+2+5, "Value Chain", fmt.Sprintf("text-anchor:left;font-size:%dpx;fill:black;font-weight:bold;font-family:serif", fontSize+2))
	s.Gend()
}
